Ingress and Terminating Gateway docs (#7710)
hashicorp/consul: website/pages/docs/agent/config-entries/ingress-gateway.mdx

---
layout: docs
page_title: 'Configuration Entry Kind: Ingress Gateway'
sidebar_title: ingress-gateway
description: >-
  The ingress-gateway config entry kind allows for configuring Ingress gateways
  with listeners that expose a set of services outside the Consul service mesh.
---

> 

Reviewer: This file should have the note that the entry is available starting in v1.8.0

# Ingress Gateway

  The `ingress-gateway` config entry kind allows for configuring Ingress gateways
  with listeners that expose a set of services outside the Consul service mesh.
  See [Ingress Gateway](/docs/connect/ingress_gateway) for more information.

## Sample Config Entries

Set up a TCP listener for a single service:

```hcl
kind = "ingress-gateway"
name = "ingress-service"
namespace = "default"

listeners = [
  {
    port = 3456
    protocol = "tcp"
    services = [
      {
        name = "db"
      }
    ]
  }
]
```

Set up a wildcard HTTP listener to proxy traffic to all available services, and
make two services available over a custom port with user-provided hosts:

```hcl
kind = "ingress-gateway"
name = "ingress-service"
namespace = "default"

listeners = [
  {
    port = 8080
    protocol = "http"
    services = [
      {
        name = "*"
      }
    ]
  },
  {
    port = 4567
    protocol = "http"
    services = [
      {
        name = "api"
        hosts ["foo.example.com"]
      },
      {
        name = "web"
        hosts ["website.example.com"]
      }
    ]
  }
]
```

## Available Fields

- `Kind` - Must be set to `ingress-gateway`

> - `Name` `(string: <required>)` - Set to the name of the service being configured.

Reviewer: ```suggestion
- `Name` `(string: <required>)` - Set to the name of the gateway being configured.
```

  - `Port` `(int: 0)` - The port that the listener should receive traffic on.

  - `Protocol` `(string: "tcp")` - The protocol associated with the listener.

  - `Services` `(array<IngressService>: <optional>)` - A list of services to be
    exposed via this listener. For "tcp" listeners, only a single service is
    allowed.

    - `Name` `(string: "")` - The name of the service that should be exposed
      through this listener. If the wildcard specifier, `*`, is provided, then
      ALL services will be exposed through the listener. This is not supported
      for listener's with protocol "tcp".

    - `Namespace` `(string: "")` - The namespace to resolve the service from
      instead of the current namespace. If empty the current namespace is
      assumed.

    - `Hosts` `(array<string>: <optional>)` - A list of hosts that specify what
      requests will match to this service. This cannot be used with a `tcp`
      listener, and cannot be specified alongside a `*` service name.

## ACLs

Configuration entries may be protected by
[ACLs](https://learn.hashicorp.com/consul/security-networking/production-acls).

> Reading a `ingress-gateway` config entry requires `service:read` on the `Name`

Reviewer 2: ```suggestion
Reading an `ingress-gateway` config entry requires `service:read` on the `Name`
```

field of the config entry.

> Creating, updating, or deleting a `ingress-gateway` config entry requires

Reviewer 2: ```suggestion
Creating, updating, or deleting an `ingress-gateway` config entry requires
```

---
layout: docs
page_title: 'Configuration Entry Kind: Ingress Gateway'
sidebar_title: ingress-gateway
description: >-

>   The ingress-gateway config entry kind allows for configuring Ingress gateways

Reviewer 3: ```suggestion
  The `ingress-gateway` config entry kind allows for configuring Ingress gateways
```

Reviewer: Done

  with listeners that expose a set of services outside the Consul service mesh.
---

# Ingress Gateway

-> **1.8.0+:** This config entry is available in Consul versions 1.8.0 and newer.

>   The `ingress-gateway` config entry kind allows for configuring Ingress gateways

Reviewer 3: ```suggestion
  The `ingress-gateway` config entry kind allows you to configure Ingress gateways
```

  with listeners that expose a set of services outside the Consul service mesh.
  See [Ingress Gateway](/docs/connect/ingress_gateway) for more information.

## Wildcard service specification

>   Ingress gateways can optionally target all services within a namespace by

Reviewer 3: ```suggestion
  Ingress gateways can optionally target all services within a Consul namespace by
```
Do you think K8s users would be confused if we didn't specify Consul? Could they think we mean k8s namespace?

Reviewer: It's possible. Changed these to "Consul namespace".

  specifying a wildcard `*` as the service name. A wildcard specifier allows
  for a single listener to route traffic to all available services on the

>   Consul mesh, differentiating between the services by their host/authority

Reviewer 3: ```suggestion
  Consul service mesh, differentiating between the services by their host/authority
```
